First set of changes intended to fix the bug

ukmo-juan-castillo opened the pull request:
Bug fix to issue #314

ukmo-juan-castillo committed:
First set of changes intended to fix the bug

ukmo-juan-castillo committed:
Small simplification

ukmo-juan-castillo commented:
These changes have already been tested in the context of issue #285 (using a manual merge between branches), and regtests are being run. Comments and possible improvements are welcome.

ukmo-juan-castillo commented:
Regtests behave as expected, changing only ww3_tp2.15, which reads input fields.
[matrixDiff.tar.gz](https://github.com/NOAA-EMC/WW3/files/6022460/matrixDiff.tar.gz)

JessicaMeixner-NOAA commented:
Can you please post the matrixCompFull.out or matrixCompSummary.out as well?

ukmo-juan-castillo commented:
No problem:

[matrixCompFull.tar.gz](https://github.com/ukmo-waves/WW3/files/6022673/matrixCompFull.tar.gz)

JessicaMeixner-NOAA commented:
Instead of running matrix.comp for each test can you run "all", that way we just have 1 file that reports on every single regression test? That's what I'm used to looking at instead of all of the individual files.   The code itself looks fine but I'm not an expert on these routines.  My only comment/suggestion would be adding more comments in ww3_prnc with the added texts for why you are adding the other parts.    If there are no changes to any of the regression tests, I see no issues with this PR.  I'll wait to run the regtests on the NCEP machines when this is pushed back to the NOAA-EMC/ww3 repo.

ukmo-juan-castillo committed:
Add comments so that the extra code is better understood

JessicaMeixner-NOAA reviewed (approved):
Again - not an expert with this code, but it all looks okay as long as there are no changes to any existing regression tests, and fixes the issue you had on the new one being added to me.  Thanks for the extra code comments! I'll run regtests when this gets pushed to noaa-emc.

ukmo-juan-castillo commented:
I will send the regtests files in a while with the format you wanted.

I wanted just to emphasize that this is a bug that was already present in the code. Why it arose now is because the density is used as a dividing factor in the code, so we generate division by zero errors. Zeroes and interpolation errors were already present for winds, currents, and water levels read from file, only that they did not generate crashes nor unreasonable results. I expect that with this fix the quality of the model output along the coastlines will improve, although not in coupled mode as the interpolation in this case is made via OASIS, or when the input files contain information for the whole grid domain and not only for wet points.

ukmo-juan-castillo commented:
[matrixCompSummary.out.txt](https://github.com/ukmo-waves/WW3/files/6023985/matrixCompSummary.out.txt)

ukmo-ccbunney commented on model/ftn/ww3_prnc.ftn:
```diff
@@ -1809,6 +1812,33 @@
+                    ! coefficients scaled so that they sum 1 - otherwise
+                    ! we will create unphysical values along the coast
+                    NORM = 0.0
+                    IF(ABS(XC(IX21(IX,IY),IY21(IX,IY))).GT.EPS)       &
```
Is it safe to assume that cells with a value of zero should not be included in the interpolation? There might be legitimate cases where a zero value is valid for a field (such as wind speed, current speed, etc).

ukmo-juan-castillo replied:
I agree that a zero value is sometimes a valid value, but it is also the value that is set by default to the land points - and by extension the value of boundary points that are not ocean points. The best way would be using this piece of code only if the grid point we are dealing with is not a boundary point. I will investigate, thanks for the useful comment.

ukmo-ccbunney replied:
>  but it is also the value that is set by default to the land points

What system is setting the land points as zero? WW3? The OASIS coupler? Is it something we have control of as it would be better if it was set to a dedicated "missing value" rather than zero?

ukmo-juan-castillo replied:
In one WW3 routine all the values read from the input forcing that contain a fill value are set to zero. I am changing the code to check for MAPSTA instead to see if this works.

ukmo-juan-castillo replied:
I have found a better solution but it is more involved - it will take me a while to sort everything out, I may create a routine so that I do not repeat code.

ukmo-ccbunney replied:
> 
> 
> In one WW3 routine all the values read from the input forcing that contain a fill value are set to zero. I am changing the code to check for MAPSTA instead to see if this works.

Ah yes - that is ringing a bell now. I think I ran into this problem with v4.18 of ww3_prnc.

ukmo-ccbunney reviewed (commented):
@ukmo-juan-castillo - just a comment regarding the test for "zero" fields.

ukmo-juan-castillo committed:
Changes to improve the code: the possibility of reading zero values from
the input is considered, and points that should not be taken into
account in the interpolation are identified by the netcdf fill value; a
subroutine is created to avoid code duplication

ukmo-juan-castillo commented:
As before, the differences are in ww3_tp2.15, but they are expected:

[matrix.tar.gz](https://github.com/ukmo-waves/WW3/files/6031471/matrix.tar.gz)

ukmo-ccbunney commented on model/ftn/ww3_prnc.ftn:
```diff
@@ -1804,23 +1802,27 @@
 !
                 IF (NFCOMP.EQ.2) THEN
 !/O3             IF ( IAPROC .EQ. NAPOUT )    WRITE (NDSO,976) ' (2) '
                     DO IY=1,NY
```
This looks like a better approach, Juan.
My only concern is that the `POINT_INTERP` subroutine is being called for every grid cell. This feels like an unnecessary overhead popping stuff on and off the call stack. Can `POINT_INTERP` take arrays instead of scalars and loop over `NX` and `NY` internally?

ukmo-juan-castillo replied:
Thanks for the comment, I thought following this approach, and I was not sure what was more efficient. In any case I guess the loss of performance will not be very important if any, but the code will be clearer too. I will implement the changes and commit when they are ready.

ukmo-juan-castillo replied:
The changes are now implemented. I tested them and they do not give exactly the same answer than the previous version of the code, but the differences are so small that can be considered rounding errors.

ukmo-ccbunney replied:
That's odd - I wouldn't have expected moving the loop into the subroutine to have changed the values at all...Perhaps it is an effect of compiler optimization?

ukmo-juan-castillo replied:
You are right, in the previous version of the code there was a mismatch between the parameters of the subroutine and the calls to it: RD21 and RD12 were exchanged, so the code was wrong. I believe the code is now right and reproduces the original behaviour, please double check.

I will send a quick test to check that this is indeed the case.

ukmo-ccbunney replied:
Great. I can't see in the code where the error was (with respect to RD21 and RD22 being in the incorrect order), but if it is now giving consistent results with the version prior to you doing the refactoring then I am happy it is good.

ukmo-juan-castillo replied:
Thanks Chris. The results before and after refactoring are the same for the density field (I was centred in fixing this), but there are differences in the wind field, and the code before refactoring seems to provide a better interpolation. I will investigate this point and modify the code until I am satisfied with the results.

ukmo-juan-castillo replied:
The code is now fixed. What was missing was calling the new interpolation routine with the right fill value for derived variables. The results of this new code are exactly the same as the code before refactoring. Thanks for all your comments, they were quite useful.

ukmo-ccbunney replied:
Nice one @ukmo-juan-castillo

ukmo-ccbunney reviewed (changes requested):
The change to handle MDI values looks sensible, Juan. I have one small refactor suggestion though.

ukmo-juan-castillo committed:
Change the new routine so that it deals with the whole set of grid
points instead of only one

ukmo-juan-castillo committed:
Fixes to properly account for the fill value when calling the new
interpolation routine
